cpp/arcticdb/storage/s3/detail-inl.hpp

```diff
@@ -44,6 +44,51 @@ namespace s3 {
         template<class It>
         using Range = folly::Range<It>;
 
+        class UnexpectedS3ErrorException : public std::exception {
+        public:
+            UnexpectedS3ErrorException(const std::string& message):message(message){}
+            const char* what() const noexcept override {
+                return message.c_str();
+            }
+        private:
+            std::string message;
+        };
+
+        inline void raise_s3_exception(const Aws::S3::S3Error& err){
+            std::string error_message;
+            // We create a more detailed error explanation in case of NETWORK_CONNECTION errors to remedy #880.
+            if (err.GetErrorType() == Aws::S3::S3Errors::NETWORK_CONNECTION){
+                error_message = fmt::format("Got an unexpected network error: {}. "
+                                            "This could be due to a connectivity issue or too many open Arctic instances. "
+                                            "Having more than one open Arctic instance is not advised, you should reuse them. "
+                                            "If you absolutely need many open Arctic instances, consider increasing `ulimit -n`.",
+                                            err.GetMessage().c_str());
```

Comment:
Does `err.GetExceptionName()` give any useful info in these cases? NETWORK_CONNECTION seems to cover so many different problems (eg the SSL problems I was looking at this week). So any extra info we can add to the message will help

Reply:
Unfortunately it's been an empty string during some of my tests.
I still agree including it is good, if at any point aws-sdk adds some more detailed errors it would be nice to have those in, so I added it.